### DISCUSSION TITLE
What is the best practice to share a massive CPU tensor over multiple processes in pytorch-lightning DDP mode (read-only + single machine)?

### DISCUSSION THREAD

**Discussion by siahuat0727** [Lightning-AI/pytorch-lightning#8611] [DDP / multi-GPU / multi-node] [answered]
Hi everyone, I wonder what is the best practice to share a massive CPU tensor over multiple processes in pytorch-lightning DDP mode (read-only + single machine)?

I think [torch.Storage.from_file](https://pytorch.org/docs/stable/storage.html?highlight=from_file#torch.DoubleStorage.from_file) with `share=True` may suit my needs, but I can’t find a way to save storage and read it as a tensor. (see [here](https://discuss.pytorch.org/t/how-to-save-a-storage-file-that-can-be-loaded-using-torch-storage-from-file/127737) for details)

I also tried to copy training data to /dev/shm ([reference](https://pytorch-lightning.readthedocs.io/en/stable/benchmarking/performance.html#preload-data-into-ram)) and run DDP with 8 GPUs, but nothing is different. The memory usage when running with 8 GPUs is the same as before, but I tested with a single process, loading the dataset may occupy more than 1 GB of memory. Am I missing something here?

For `torch.shared_memory`, how should I pass the same reference to all processes in pytorch-lightning pure DDP mode?

Thank you.

**Comment by tchaton** [comment]
Dear @siahuat0727,

Lightning doesn't support shared tensors yet, but there is some work being done around it. 
You can track this issue: https://github.com/PyTorchLightning/pytorch-lightning/issues/8230 and help if you feel like :)

Best,
TC

**Reply by siahuat0727** [reply]
Got it! Thank you :)

**Comment by siahuat0727** [comment, marked as answer]
I found that `torch.Storage.from_file` suits my needs and it can reduce the memory usage in my Lightning DDP program.
For the way to create a storage file, see [here](https://discuss.pytorch.org/t/how-to-save-a-storage-file-that-can-be-loaded-using-torch-storage-from-file/127737).
